Changes to build against Fast DDS 3.0

MiguelCompany opened the pull request:
This PR brings in the necessary changes to build against Fast DDS 3.0.x

It has been reviewed by the eProsima Team in eProsima#1

Part of ros2/ros2#1598

richiware committed:
Support for `Fast DDS 3.0`

* Fast DDS 3.0: first commit

Signed-off-by: Ricardo González Moreno <ricardo@richiware.dev>

* Refs #21322: update methods to

Signed-off-by: Mario Dominguez <mariodominguez@eprosima.com>

* Refs #21322: update identifier to

Signed-off-by: Mario Dominguez <mariodominguez@eprosima.com>

* Refs #21322: Avoid compilation warnings

Signed-off-by: Mario Dominguez <mariodominguez@eprosima.com>

---------

Signed-off-by: Ricardo González Moreno <ricardo@richiware.dev>
Signed-off-by: Mario Dominguez <mariodominguez@eprosima.com>
Co-authored-by: Mario Dominguez <mariodominguez@eprosima.com>

MiguelCompany committed:
Do not require cmake modules.

Signed-off-by: Miguel Company <MiguelCompany@eprosima.com>

MiguelCompany committed:
Remove dependency on fastrtps_cmake_module.

Signed-off-by: Miguel Company <MiguelCompany@eprosima.com>

MiguelCompany committed:
Fix dll linkage and keep header guards.

Miguel Company <miguelcompany@eprosima.com>

richiware committed:
Fix windows warnings

Signed-off-by: Ricardo González <ricardo@richiware.dev>

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+    RCUTILS_SET_ERROR_MSG("Could not init new data");
+    return RCUTILS_RET_ERROR;
+  }
+
```
I'm going to suggest that we use an `rcpputils::scope_exit` here to `delete loaned_data_impl_handle`.  That will add a new dependency, but I think it is more robust for us adding in other error paths in the future.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+  FASTDDS_CHECK_RET_FOR_NOT_OK_WITH_MSG(
+    data_handle->ref_type->return_loaned_value(inner_data_handle->ref_type),
+    "Could not return loaned value");
+  //TODO(richiware) remove structure?
```
```suggestion
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+  }
+
+  std::string tmp_name = data_handle->ref_type->type()->get_name().to_string();
   *name = rcutils_strdup(tmp_name.c_str(), rcutils_get_default_allocator());
```
I know you didn't add this, but this should check for a NULL return value:

```suggestion
  *name = rcutils_strdup(tmp_name.c_str(), rcutils_get_default_allocator());
  if (*name == nullptr) {
    RCUTILS_SET_ERROR_MSG("Failed to duplicate name for data name");
    return RCUTILS_RET_BAD_ALLOC;
  }
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+  }
+
+  auto data_impl_handle =
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_data_impl();
```
Use an `rcpputils::scope_exit` to free `data_impl_handle` on errors.

clalancette replied:
Also, check for NULL `data_impl_handle` after the `new`.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  auto data_impl_handle =
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_data_impl();
```
Use an `rcpputils::scope_exit` to free the `data_impl_handle` on error.

clalancette replied:
Also, check for NULL after the `new`.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  auto data_impl_handle =
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_data_impl();
```
Use an `rcpputils::scope_exit` to free `data_impl_handle` on error.

clalancette replied:
Also, check for NULL after the `new`.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -559,445 +743,616 @@ fastrtps__dynamic_data_get_bounded_string_value(
+  (void)serialization_support_impl;
+  (void)data_impl;
+  (void)out_id;
+  //TODO(richiware)
```
Does this need to be implemented?

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
Should we rename this file to `fastdds_dynamic_data.cpp`?

richiware replied:
We commented this when we started the implementation changes, and we thought this can be done in a separate PR.

clalancette commented on src/detail/fastrtps_dynamic_data.hpp:
Should we rename this file to `fastdds_dynamic_data.hpp`?

richiware replied:
We commented this when we started the implementation changes, and we thought this can be done in a separate PR.

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
Should we rename this file to `fastdds_dynamic_type.cpp`?

richiware replied:
We commented this when we started the implementation changes, and we thought this can be done in a separate PR.

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -35,105 +34,122 @@
-using eprosima::fastrtps::types::DynamicTypeBuilder;
-using eprosima::fastrtps::types::DynamicTypeBuilder_ptr;
-using eprosima::fastrtps::types::TypeDescriptor;
+// using eprosima::fastdds::types::DynamicType;  // Conflicts in this scope for some reason...
```
```suggestion
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -35,105 +34,122 @@
+  // We must replace "/" with "::" in type names
+  std::string name_string = fastdds__replace_string(std::string(name, name_length), "/", "::");
+
+  auto type_builder_handle =
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_type_builder_impl();
```
We should use an `rcpputils::scope_exit` to cleanup on error here.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -35,105 +34,122 @@
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type builder");
+    return RCUTILS_RET_ERROR;
+  }
+  auto other_type_builder_handle =
```
We should use an `rcpputils::scope_exit` here to cleanup `type_builder_handler` on error.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -35,105 +34,122 @@
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_type_builder_impl();
   if (!type_builder_handle) {
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type builder");
+    return RCUTILS_RET_ERROR;
```
```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
+    return RCUTILS_RET_ERROR;
+  }
+  type_builder_handle->ref_type.reset();
+  //TODO(richiware) delete?
```
```suggestion
  delete type_builder_handle;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
+    static_cast<fastdds__serialization_support_impl_handle_t *>(serialization_support_impl->handle)
+    ->data_factory_->delete_data(data_handle->ref_type),
+    "Could not fini data");
+  //TODO(richiware) delete structure?
```
```suggestion
  delete data_handle;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
+  auto type_handle = new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_type_impl();
+  if (!type_handle) {
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type");
+    return RCUTILS_RET_ERROR;
```
```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type");
+    return RCUTILS_RET_ERROR;
+  }
+
```
Use an `rcpputils::scope_exit` to free `type_handle` on error.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type");
     return RCUTILS_RET_ERROR;
   }
 
```
Use an `rcpputils::scope_exit` to delete `type_impl_out_handle` on error.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
+  FASTDDS_CHECK_RET_FOR_NOT_OK_WITH_MSG(
+    fastdds_impl->type_factory_->delete_type(type_handle->ref_type),
+    "Could not fini type");
+  //TODO(richiware) delete structure?
```
```suggestion
  delete type_handle;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
+  (void)name;
+  (void)name_length;
+  return RCUTILS_RET_OK;
+  // TODO(richiware)
```
Does this need to be implemented?

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.hpp:
Should we rename this file to `fastdds_dynamic_type.hpp`?

clalancette commented on src/detail/fastrtps_serialization_support.cpp:
```diff
@@ -22,31 +21,31 @@
-  allocator.deallocate(serialization_support_impl->handle, allocator.state);
-  return RCUTILS_RET_OK;
+    allocator.deallocate(serialization_support_impl->handle, allocator.state);
+    return RCUTILS_RET_OK;
```
Do we need to change all of this indentation?  Or can we just leave it as it was?

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_serialization_support.cpp:
```diff
@@ -22,31 +21,31 @@
-  static_cast<void>(serialization_support_interface);
-  return RCUTILS_RET_OK;
+    static_cast<void>(serialization_support_interface);
+    return RCUTILS_RET_OK;
```
Same question here.

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_serialization_support.hpp:
Should we rename this file to `fastdds_serialization_support.hpp`?

richiware replied:
We commented this when we started the implementation changes, and we thought this can be done in a separate PR.

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +38,418 @@
-    fastrtps__size_t_to_uint32_t(id));
   if (!loaned_data_impl_handle) {
+    RCUTILS_SET_ERROR_MSG("Could not init new data");
+    return RCUTILS_RET_ERROR;
```
```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -35,105 +34,122 @@
   if (!type_builder_handle) {
     RCUTILS_SET_ERROR_MSG("Could not init new struct type builder");
-    return RCUTILS_RET_BAD_ALLOC;
+    return RCUTILS_RET_ERROR;
```
I think that `RCUTILS_RET_BAD_ALLOC` is correct here:

```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

richiware replied:
Done in 2ec21b3

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
   if (!type_impl_out_handle) {
-    RCUTILS_SET_ERROR_MSG("Could not clone struct type");
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type");
     return RCUTILS_RET_ERROR;
```
I know that this is existing code, but this should be a BAD_ALLOC:

```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

richiware replied:
Done in 2ec21b3

richiware committed:
Apply suggestions

Signed-off-by: Ricardo González Moreno <ricardo@richiware.dev>

fujitatomoya commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +40,472 @@
-}
+  (void)serialization_support_impl;
 
+  auto data_handle =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_data_impl *>(data_impl->handle);
+  if (!data_handle || !data_handle->ref_type) {
+    RCUTILS_SET_ERROR_MSG("Could not get handle to data impl");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
```
i would create the macro to check this, because there are many places that do the same check.

clalancette replied:
I'm not personally a huge fan of using macros for this; I think they tend to be hard to read later on.  But I don't feel strongly about it, so whatever @richiware decides is OK with me.

fujitatomoya commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +40,472 @@
+    ->data_factory_->create_data(type_builder_handle->ref_type->build());
+  if (!data_impl_handle->ref_type) {
     RCUTILS_SET_ERROR_MSG("Could not init dynamic data from dynamic type builder");
+    delete data_impl_handle;
```
is this going to be a double-free situation (undefined behavior, could be nothing) once cleanup_data_impl_handle  is out of scope?

clalancette replied:
Yes, agreed that this is no longer needed because of the `rcpputils::make_scope_exit`:

```suggestion
```

fujitatomoya commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +40,472 @@
+    ->data_factory_->create_data(type_handle->ref_type);
+  if (!data_impl_handle->ref_type) {
     RCUTILS_SET_ERROR_MSG("Could not init dynamic data from dynamic type");
+    delete data_impl_handle;
```
same here, this could be double-free? (and other similar places with `rcpputils::make_scope_exit`)

clalancette replied:
Agreed.

```suggestion
```

fujitatomoya reviewed (commented):
a couple of minor comments.

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -36,285 +40,472 @@
+  data_impl_handle->ref_type = other_data_handle->ref_type->clone();
   if (!data_impl_handle) {
     RCUTILS_SET_ERROR_MSG("Could not clone struct type builder");
+    delete data_impl_handle;
```
No longer needed.

```suggestion
```

clalancette commented on src/detail/fastrtps_dynamic_data.cpp:
```diff
@@ -324,9 +515,13 @@ fastrtps__dynamic_data_serialize(
+      data_length));
+  bool success = m_type->serialize(
+    &data_handle->ref_type, *payload,
+    eprosima::fastdds::dds::DataRepresentationId_t::XCDR_DATA_REPRESENTATION);                                                                    // Serialize into payload
```
I know we aren't enforcing linting here, but I think we can shrink the whitespace here a lot:

```suggestion
    eprosima::fastdds::dds::DataRepresentationId_t::XCDR_DATA_REPRESENTATION);  // Serialize into payload
```

clalancette commented on src/detail/fastrtps_dynamic_data.hpp:
```diff
@@ -31,660 +31,714 @@
   const rosidl_dynamic_typesupport_dynamic_data_impl_t * data_impl,
   const rosidl_dynamic_typesupport_dynamic_data_impl_t * other_data_impl,
-  bool * equals);  // OUT
+  bool * equals);      // OUT
```
The whitespace change here seems unnecessary.  Same goes for all of the ones below.

(this will help reduce the size of this review a bit)

clalancette commented on src/detail/utils.cpp:
```diff
@@ -24,7 +24,8 @@
 
 uint32_t
-fastrtps__size_t_to_uint32_t(size_t in)
+fastdds__size_t_to_uint32_t(
+  size_t in)
```
This seems like an unnecessary whitespace change:

```suggestion
fastdds__size_t_to_uint32_t(size_t in)
```

(the same goes for most of the functions in this file)

clalancette reviewed (changes requested):
I've left a few more minor comments.  I think this is pretty close to being ready.

MiguelCompany committed:
Avoid deleting things twice

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Signed-off-by: Miguel Company <miguelcompany@eprosima.com>

MiguelCompany committed:
Remove whitespace changes

Signed-off-by: Miguel Company <miguelcompany@eprosima.com>

MiguelCompany commented:
@clalancette I have rebased this, and also removed the unnecessary `delete` calls, as well as the superfluous whitespace changes.

I think it is ready for another review.

clalancette commented on src/detail/fastrtps_dynamic_type.cpp:
```diff
@@ -35,1019 +37,1411 @@
-
-rcutils_ret_t
-fastrtps__dynamic_type_builder_clone(
+rcutils_ret_tfastdds__dynamic_type_builder_clone(
```
```suggestion
rcutils_ret_t
fastdds__dynamic_type_builder_clone(
```

MiguelCompany committed:
Apply suggestion

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Signed-off-by: Miguel Company <miguelcompany@eprosima.com>

clalancette commented:
CI for this is in https://github.com/ros2/ros2/pull/1598#issuecomment-2607298187
